Intermittent Sleuth failures on GraphQL service

We can see the following stack trace when the GraphQL service is running:
```
Exception in thread "clientInboundChannel-1" org.springframework.messaging.MessageDeliveryException: Failed to handle GenericMessage [payload=GraphQLMessage [payload={query=subscription($appName: String!, $eventTypes: [EngineEventType!]) {   engineEvents(appName: [$appName], eventType: $eventTypes) {     processInstanceId     processDefinitionId     entity     eventType   } }, variables={appName=default-app, eventTypes=[MESSAGE_SENT, MESSAGE_WAITING, MESSAGE_RECEIVED]}}, id=1, type=start], headers={simpMessageType=MESSAGE, nativeHeaders={}, simpSessionAttributes={HTTP.SESSION.ID=C840AA5029CCCD42F5B0496CFF2D9293, SPRING_SECURITY_CONTEXT=SecurityContextImpl [Authentication=KeycloakAuthenticationToken [Principal=5f682999-d11d-4a42-bc42-86b7e675bb76, Credentials=[PROTECTED], Authenticated=true, Details=org.keycloak.adapters.springsecurity.account.SimpleKeycloakAccount@7c857768, Granted Authorities=[ROLE_ACTIVITI_ADMIN, ROLE_offline_access, ROLE_ACTIVITI_USER, ROLE_uma_authorization]]]}, simpUser=KeycloakAuthenticationToken [Principal=5f682999-d11d-4a42-bc42-86b7e675bb76, Credentials=[PROTECTED], Authenticated=true, Details=org.keycloak.adapters.springsecurity.account.SimpleKeycloakAccount@72429688, Granted Authorities=[ROLE_ACTIVITI_ADMIN, ROLE_offline_access, ROLE_ACTIVITI_USER, ROLE_uma_authorization]], graphQLMessageType=start, simpSessionId=d9de796c-2134-150d-1635-65694c6019b5, simpDestination=/ws/graphql}] to org.springframework.messaging.support.ExecutorSubscribableChannel$SendTask@767d1c18 in org.activiti.cloud.services.notifications.graphql.ws.transport.GraphQLBrokerMessageHandler@1d215a52; nested exception is java.lang.IllegalStateException: Already immutable, failedMessage=GenericMessage [payload=GraphQLMessage [payload={query=subscription($appName: String!, $eventTypes: [EngineEventType!]) {   engineEvents(appName: [$appName], eventType: $eventTypes) {     processInstanceId     processDefinitionId     entity     eventType   } }, variables={appName=default-app, eventTypes=[MESSAGE_SENT, MESSAGE_WAITING, MESSAGE_RECEIVED]}}, id=1, type=start], headers={simpMessageType=MESSAGE, nativeHeaders={}, simpSessionAttributes={HTTP.SESSION.ID=C840AA5029CCCD42F5B0496CFF2D9293, SPRING_SECURITY_CONTEXT=SecurityContextImpl [Authentication=KeycloakAuthenticationToken [Principal=5f682999-d11d-4a42-bc42-86b7e675bb76, Credentials=[PROTECTED], Authenticated=true, Details=org.keycloak.adapters.springsecurity.account.SimpleKeycloakAccount@7c857768, Granted Authorities=[ROLE_ACTIVITI_ADMIN, ROLE_offline_access, ROLE_ACTIVITI_USER, ROLE_uma_authorization]]]}, simpUser=KeycloakAuthenticationToken [Principal=5f682999-d11d-4a42-bc42-86b7e675bb76, Credentials=[PROTECTED], Authenticated=true, Details=org.keycloak.adapters.springsecurity.account.SimpleKeycloakAccount@72429688, Granted Authorities=[ROLE_ACTIVITI_ADMIN, ROLE_offline_access, ROLE_ACTIVITI_USER, ROLE_uma_authorization]], graphQLMessageType=start, simpSessionId=d9de796c-2134-150d-1635-65694c6019b5, simpDestination=/ws/graphql}]
	at org.springframework.messaging.support.ExecutorSubscribableChannel$SendTask.run(ExecutorSubscribableChannel.java:153)
	at org.springframework.cloud.sleuth.instrument.async.TraceRunnable.run(TraceRunnable.java:63)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Already immutable
	at org.springframework.util.Assert.state(Assert.java:76)
	at org.springframework.messaging.support.NativeMessageHeaderAccessor.removeNativeHeader(NativeMessageHeaderAccessor.java:277)
	at org.springframework.cloud.sleuth.instrument.messaging.MessageHeaderPropagatorSetter.removeAnyTraceHeaders(MessageHeaderPropagatorSetter.java:62)
	at org.springframework.cloud.sleuth.instrument.messaging.TracingChannelInterceptor.beforeHandle(TracingChannelInterceptor.java:324)
	at org.springframework.messaging.support.ExecutorSubscribableChannel$SendTask.applyBeforeHandle(ExecutorSubscribableChannel.java:167)
	at org.springframework.messaging.support.ExecutorSubscribableChannel$SendTask.run(ExecutorSubscribableChannel.java:140)
	... 4 more
```

ActivitiGraphQLStarterIT also fails intermittently and when this happens we can see the stack trace above.

--- a/activiti-cloud-notifications-graphql-service/services/ws/src/main/resources/META-INF/graphql-ws.properties
+++ b/activiti-cloud-notifications-graphql-service/services/ws/src/main/resources/META-INF/graphql-ws.properties
@@ -9,3 +9,5 @@ spring.activiti.cloud.services.notifications.graphql.ws.endpoint=/ws/graphql
 spring.activiti.cloud.services.notifications.graphql.ws.allowed-origins=*
 spring.activiti.cloud.services.notifications.graphql.ws.buffer-count=50
 spring.activiti.cloud.services.notifications.graphql.ws.buffer-timespan-ms=1000
+
+spring.sleuth.integration.patterns=[!hystrixStreamOutput*, ,!channel,!clientInboundChannel,!clientOutboundChannel,!brokerChannel]

--- a/activiti-cloud-notifications-graphql-service/services/ws/src/main/resources/META-INF/graphql-ws.properties
+++ b/activiti-cloud-notifications-graphql-service/services/ws/src/main/resources/META-INF/graphql-ws.properties
@@ -10,4 +10,4 @@ spring.activiti.cloud.services.notifications.graphql.ws.allowed-origins=*
 spring.activiti.cloud.services.notifications.graphql.ws.buffer-count=50
 spring.activiti.cloud.services.notifications.graphql.ws.buffer-timespan-ms=1000
 
-spring.sleuth.integration.patterns=[!hystrixStreamOutput*, ,!channel,!clientInboundChannel,!clientOutboundChannel,!brokerChannel]
+spring.sleuth.integration.patterns=!hystrixStreamOutput*,!channel,!clientInboundChannel*,!clientOutboundChannel*,!brokerChannel*,*

--- a/activiti-cloud-notifications-graphql-service/starter/src/main/resources/metadata.properties
+++ b/activiti-cloud-notifications-graphql-service/starter/src/main/resources/metadata.properties
@@ -1,3 +1,6 @@
 activiti.cloud.service.type=notifications-graphql
 
 spring.jpa.open-in-view=false
+
+spring.sleuth.integration.websockets.enabled=false
+spring.sleuth.integration.enabled=false

--- a/activiti-cloud-notifications-graphql-service/starter/src/main/resources/metadata.properties
+++ b/activiti-cloud-notifications-graphql-service/starter/src/main/resources/metadata.properties
@@ -1,6 +1,3 @@
 activiti.cloud.service.type=notifications-graphql
 
 spring.jpa.open-in-view=false
-
-spring.sleuth.integration.websockets.enabled=false
-spring.sleuth.integration.enabled=false

--- a/activiti-cloud-notifications-graphql-service/starter/src/test/java/org/activiti/cloud/notifications/graphql/starter/ActivitiGraphQLStarterIT.java
+++ b/activiti-cloud-notifications-graphql-service/starter/src/test/java/org/activiti/cloud/notifications/graphql/starter/ActivitiGraphQLStarterIT.java
@@ -172,7 +172,15 @@ public class ActivitiGraphQLStarterIT {
                             .log("client-send")
                             .subscribe();
 
-                    return i.receive().asString();
+                    return i.aggregateFrames()
+                            .receive()
+                            .asString()
+                            .doOnCancel(() -> {
+                                // Let's close websocket and complete data processor
+                                o.sendClose()
+                                 .doOnTerminate(output::onComplete)
+                                 .subscribe();
+                            });
                 })
                 .log("client-received")
                 .take(2)
@@ -270,16 +278,23 @@ public class ActivitiGraphQLStarterIT {
                     .log("start")
                     .subscribe();
 
-            return i.receive()
-                           .asString()
-                           .log("data")
-                           .take(1)
-                           .doOnSubscribe(s -> producerChannel.output()
-                                                       .send(MessageBuilder.withPayload(Arrays.array(event1, event2))
-                                                                     .setHeader("routingKey", "eventProducer")
-                                                                     .build()))
-                           .delaySubscription(Duration.ofSeconds(1))
-                           .subscribeWith(data);
+            return i.aggregateFrames()
+                    .receive()
+                    .asString()
+                    .log("data")
+                    .take(1)
+                    .doOnSubscribe(s -> producerChannel.output()
+                                               .send(MessageBuilder.withPayload(Arrays.array(event1, event2))
+                                                             .setHeader("routingKey", "eventProducer")
+                                                             .build()))
+                    .delaySubscription(Duration.ofSeconds(1))
+                    .doOnCancel(() -> {
+                        // Let's close websocket and complete data processor
+                        o.sendClose()
+                         .doOnTerminate(data::onComplete)
+                         .subscribe();
+                    })
+                    .subscribeWith(data);
         }) // stop subscription
                 .collectList()
                 .subscribe();
@@ -362,16 +377,23 @@ public class ActivitiGraphQLStarterIT {
                     .log("start")
                     .subscribe();
 
-            return i.receive()
-                           .asString()
-                           .log("data")
-                           .take(1)
-                           .doOnSubscribe(s -> producerChannel.output()
-                                                       .send(MessageBuilder.withPayload(Arrays.array(event1))
-                                                                     .setHeader("routingKey", "eventProducer")
-                                                                     .build()))
-                           .delaySubscription(Duration.ofSeconds(1))
-                           .subscribeWith(data);
+            return i.aggregateFrames()
+                    .receive()
+                    .asString()
+                    .log("data")
+                    .take(1)
+                    .doOnSubscribe(s -> producerChannel.output()
+                                               .send(MessageBuilder.withPayload(Arrays.array(event1))
+                                                             .setHeader("routingKey", "eventProducer")
+                                                             .build()))
+                    .delaySubscription(Duration.ofSeconds(1))
+                    .doOnCancel(() -> {
+                       // Let's close websocket and complete data processor
+                        o.sendClose()
+                         .doOnTerminate(data::onComplete)
+                         .subscribe();
+                    })
+                    .subscribeWith(data);
         }) // stop subscription
                 .collectList()
                 .subscribe();
@@ -456,16 +478,23 @@ public class ActivitiGraphQLStarterIT {
                     .log("start")
                     .subscribe();
 
-            return i.receive()
-                           .asString()
-                           .log("data")
-                           .take(1)
-                           .doOnSubscribe(s -> producerChannel.output()
-                                                       .send(MessageBuilder.withPayload(Arrays.array(event1))
-                                                                     .setHeader("routingKey", "eventProducer")
-                                                                     .build()))
-                           .delaySubscription(Duration.ofSeconds(1))
-                           .subscribeWith(data);
+            return i.aggregateFrames()
+                    .receive()
+                    .asString()
+                    .log("data")
+                    .take(1)
+                    .doOnSubscribe(s -> producerChannel.output()
+                                               .send(MessageBuilder.withPayload(Arrays.array(event1))
+                                                             .setHeader("routingKey", "eventProducer")
+                                                             .build()))
+                    .delaySubscription(Duration.ofSeconds(1))
+                    .doOnCancel(() -> {
+                        // Let's close websocket and complete data processor
+                        o.sendClose()
+                         .doOnTerminate(data::onComplete)
+                         .subscribe();
+                    })
+                    .subscribeWith(data);
         }) // stop subscription
                 .collectList()
                 .subscribe();
@@ -554,16 +583,23 @@ public class ActivitiGraphQLStarterIT {
                     .log("start")
                     .subscribe();
 
-            return i.receive()
-                           .asString()
-                           .log("data")
-                           .timeout(Duration.ofSeconds(2))
-                           .doOnSubscribe(s -> producerChannel.output()
-                                                       .send(MessageBuilder.withPayload(Arrays.array(event1))
-                                                                     .setHeader("routingKey", "eventProducer")
-                                                                     .build()))
-                           .delaySubscription(Duration.ofSeconds(1))
-                           .subscribeWith(data);
+            return i.aggregateFrames()
+                    .receive()
+                    .asString()
+                    .log("data")
+                    .timeout(Duration.ofSeconds(2))
+                    .doOnSubscribe(s -> producerChannel.output()
+                                               .send(MessageBuilder.withPayload(Arrays.array(event1))
+                                                             .setHeader("routingKey", "eventProducer")
+                                                             .build()))
+                    .delaySubscription(Duration.ofSeconds(1))
+                    .doOnCancel(() -> {
+                        // Let's close websocket and complete data processor
+                        o.sendClose()
+                         .doOnTerminate(data::onComplete)
+                         .subscribe();
+                    })
+                    .subscribeWith(data);
         }) // stop subscription
                 .collectList()
                 .subscribe();
@@ -739,17 +775,24 @@ public class ActivitiGraphQLStarterIT {
                     .log("start")
                     .subscribe();
 
-            return i.receive()
-                           .asString()
-                           .log("data")
-                           .take(1)
-                           .doOnSubscribe(s -> producerChannel.output()
-                                                       .send(MessageBuilder
-                                                                     .withPayload(Arrays.array(event1, event2, event3, event4, event5, event6))
-                                                                     .setHeader("routingKey", "eventProducer")
-                                                                     .build()))
-                           .delaySubscription(Duration.ofSeconds(1))
-                           .subscribeWith(data);
+            return i.aggregateFrames()
+                    .receive()
+                    .asString()
+                    .log("data")
+                    .take(1)
+                    .doOnSubscribe(s -> producerChannel.output()
+                                               .send(MessageBuilder
+                                                             .withPayload(Arrays.array(event1, event2, event3, event4, event5, event6))
+                                                             .setHeader("routingKey", "eventProducer")
+                                                             .build()))
+                    .delaySubscription(Duration.ofSeconds(1))
+                    .doOnCancel(() -> {
+                       // Let's close websocket and complete data processor
+                        o.sendClose()
+                         .doOnTerminate(data::onComplete)
+                         .subscribe();
+                    })
+                    .subscribeWith(data);
         }) // stop subscription
                 .collectList()
                 .subscribe();
@@ -905,19 +948,26 @@ public class ActivitiGraphQLStarterIT {
                     .log("start")
                     .subscribe();
 
-            return i.receive()
-                           .asString()
-                           .log("data")
-                           .take(1)
-                           .doOnSubscribe(s -> producerChannel.output()
-                                                       .send(MessageBuilder.withPayload(Arrays.array(event1, event2, event3))
-                                                                     .setHeader("routingKey", "eventProducer")
-                                                                     .build()))
-                           .delaySubscription(Duration.ofSeconds(1))
-                           .subscribeWith(data);
+            return i.aggregateFrames()
+                    .receive()
+                    .asString()
+                    .log("data")
+                    .take(1)
+                    .doOnSubscribe(s -> producerChannel.output()
+                                               .send(MessageBuilder.withPayload(Arrays.array(event1, event2, event3))
+                                                             .setHeader("routingKey", "eventProducer")
+                                                             .build()))
+                    .delaySubscription(Duration.ofSeconds(1))
+                    .doOnCancel(() -> {
+                        // Let's close websocket and complete data processor
+                        o.sendClose()
+                         .doOnTerminate(data::onComplete)
+                         .subscribe();
+                    })
+                    .subscribeWith(data);
         }) // stop subscription
-                .collectList()
-                .subscribe();
+        .collectList()
+        .subscribe();
 
         // then
         Map<String, Object> message = Maps.of("data",
@@ -981,7 +1031,15 @@ public class ActivitiGraphQLStarterIT {
                             .log("client-send")
                             .subscribe();
 
-                    return i.receive().asString();
+                    return i.aggregateFrames()
+                            .receive()
+                            .asString()
+                            .doOnCancel(() -> {
+                                // Let's close websocket and complete data processor
+                                o.sendClose()
+                                 .doOnTerminate(output::onComplete)
+                                 .subscribe();
+                            });
                 })
                 .log("client-received")
                 .take(2)
@@ -1002,7 +1060,7 @@ public class ActivitiGraphQLStarterIT {
                 .expectNext(ackMessage)
                 .expectNext(kaMessage)
                 .expectComplete()
-                .verify(TIMEOUT);        
+                .verify(TIMEOUT);
     }
 
     @Test
@@ -1024,7 +1082,15 @@ public class ActivitiGraphQLStarterIT {
                             .log("client-send")
                             .subscribe();
 
-                    return i.receive().asString();
+                    return i.aggregateFrames()
+                            .receive()
+                            .asString()
+                            .doOnCancel(() -> {
+                                // Let's close websocket and complete data processor
+                                o.sendClose()
+                                 .doOnTerminate(output::onComplete)
+                                 .subscribe();
+                            });
                 })
                 .log("client-received")
                 .take(1)

--- a/activiti-cloud-notifications-graphql-service/starter/src/test/java/org/activiti/cloud/notifications/graphql/starter/ActivitiGraphQLStarterIT.java
+++ b/activiti-cloud-notifications-graphql-service/starter/src/test/java/org/activiti/cloud/notifications/graphql/starter/ActivitiGraphQLStarterIT.java
@@ -92,6 +92,7 @@ import reactor.core.publisher.ReplayProcessor;
 import reactor.netty.http.client.HttpClient;
 import reactor.netty.http.client.HttpClient.WebsocketSender;
 import reactor.netty.http.client.WebsocketClientSpec;
+import reactor.netty.http.websocket.WebsocketOutbound;
 import reactor.test.StepVerifier;
 
 @SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
@@ -176,10 +177,7 @@ public class ActivitiGraphQLStarterIT {
                             .receive()
                             .asString()
                             .doOnCancel(() -> {
-                                // Let's close websocket and complete data processor
-                                o.sendClose()
-                                 .doOnTerminate(output::onComplete)
-                                 .subscribe();
+                                closeWebSocketAnCompleteDataProcessor(output, o);
                             });
                 })
                 .log("client-received")
@@ -289,10 +287,7 @@ public class ActivitiGraphQLStarterIT {
                                                              .build()))
                     .delaySubscription(Duration.ofSeconds(1))
                     .doOnCancel(() -> {
-                        // Let's close websocket and complete data processor
-                        o.sendClose()
-                         .doOnTerminate(data::onComplete)
-                         .subscribe();
+                        closeWebSocketAnCompleteDataProcessor(data, o);
                     })
                     .subscribeWith(data);
         }) // stop subscription
@@ -319,6 +314,13 @@ public class ActivitiGraphQLStarterIT {
                 .expectNext(dataMessage)
                 .expectComplete()
                 .verify(TIMEOUT);
+    }
+
+    private void closeWebSocketAnCompleteDataProcessor(ReplayProcessor<String> data,
+        WebsocketOutbound webSocketOutbound) {
+        webSocketOutbound.sendClose()
+            .doOnTerminate(data::onComplete)
+            .subscribe();
     }
 
     @Test
@@ -388,10 +390,7 @@ public class ActivitiGraphQLStarterIT {
                                                              .build()))
                     .delaySubscription(Duration.ofSeconds(1))
                     .doOnCancel(() -> {
-                       // Let's close websocket and complete data processor
-                        o.sendClose()
-                         .doOnTerminate(data::onComplete)
-                         .subscribe();
+                        closeWebSocketAnCompleteDataProcessor(data, o);
                     })
                     .subscribeWith(data);
         }) // stop subscription
@@ -489,10 +488,7 @@ public class ActivitiGraphQLStarterIT {
                                                              .build()))
                     .delaySubscription(Duration.ofSeconds(1))
                     .doOnCancel(() -> {
-                        // Let's close websocket and complete data processor
-                        o.sendClose()
-                         .doOnTerminate(data::onComplete)
-                         .subscribe();
+                        closeWebSocketAnCompleteDataProcessor(data, o);
                     })
                     .subscribeWith(data);
         }) // stop subscription
@@ -594,10 +590,7 @@ public class ActivitiGraphQLStarterIT {
                                                              .build()))
                     .delaySubscription(Duration.ofSeconds(1))
                     .doOnCancel(() -> {
-                        // Let's close websocket and complete data processor
-                        o.sendClose()
-                         .doOnTerminate(data::onComplete)
-                         .subscribe();
+                        closeWebSocketAnCompleteDataProcessor(data, o);
                     })
                     .subscribeWith(data);
         }) // stop subscription
@@ -787,10 +780,7 @@ public class ActivitiGraphQLStarterIT {
                                                              .build()))
                     .delaySubscription(Duration.ofSeconds(1))
                     .doOnCancel(() -> {
-                       // Let's close websocket and complete data processor
-                        o.sendClose()
-                         .doOnTerminate(data::onComplete)
-                         .subscribe();
+                        closeWebSocketAnCompleteDataProcessor(data, o);
                     })
                     .subscribeWith(data);
         }) // stop subscription
@@ -959,10 +949,7 @@ public class ActivitiGraphQLStarterIT {
                                                              .build()))
                     .delaySubscription(Duration.ofSeconds(1))
                     .doOnCancel(() -> {
-                        // Let's close websocket and complete data processor
-                        o.sendClose()
-                         .doOnTerminate(data::onComplete)
-                         .subscribe();
+                        closeWebSocketAnCompleteDataProcessor(data, o);
                     })
                     .subscribeWith(data);
         }) // stop subscription
@@ -1035,10 +1022,7 @@ public class ActivitiGraphQLStarterIT {
                             .receive()
                             .asString()
                             .doOnCancel(() -> {
-                                // Let's close websocket and complete data processor
-                                o.sendClose()
-                                 .doOnTerminate(output::onComplete)
-                                 .subscribe();
+                                closeWebSocketAnCompleteDataProcessor(output, o);
                             });
                 })
                 .log("client-received")
@@ -1086,10 +1070,7 @@ public class ActivitiGraphQLStarterIT {
                             .receive()
                             .asString()
                             .doOnCancel(() -> {
-                                // Let's close websocket and complete data processor
-                                o.sendClose()
-                                 .doOnTerminate(output::onComplete)
-                                 .subscribe();
+                                closeWebSocketAnCompleteDataProcessor(output, o);
                             });
                 })
                 .log("client-received")

